R/case-variables.R

```diff
@@ -0,0 +1,169 @@
+#' Validate case statements for case variables
+#'
+#' List elements:
+#' id an integer to use for this category when a case variable is made
+#' (default: none, one will automatically be assigned when the case variable
+#' is made)
+#' name a charcater identifier for this case
+#' expression `CrunchLogicalExpr` which sets the conditions for the case
+#' numeric_value a numeric which is the value this case should take on
+#' (useful when made into a case variable)
+#' missing a logical indicating if this case should be treated as missing
+#'
+#' @keywords internal
+ensureValidCase <- function(case) {
+    if (!is.list(case)) {
+        halt("A case must be a list")
+    }
+    wrong_case_names <- setdiff(names(case), c("id", "name", "expression",
+                                               "numeric_value", "missing"))
+    if (length(wrong_case_names) > 0) {
+        halt("each case must have at most an id, name, expression, ",
+             "numeric_value, and missing element. The errant arguments were: ",
+             serialPaste(wrong_case_names))
+    }
+    if (!is.null(case$id) && !is.whole(case$id)) {
+        halt("a case's id must be an integer")
+    }
+    if (!is.character(case$name)) {
+        halt("a case's name must be a character")
+    }
+    if (class(case$expression) != "CrunchLogicalExpr") {
+        halt("a case's expression must be a CrunchLogicalExpr")
+    }
+    if (is.null(case$numeric_value)) {
+        case['numeric_value'] <- list(NULL)
+    } else if (!is.numeric(case$numeric_value)) {
+        halt("a case's numeric_value must be a numeric")
+    }
+    if (is.null(case$missing)) {
+        case['missing'] <- FALSE
+    } else if (!is.logical(case$missing)) {
+        halt("a case's missing must be a logical")
+    }
+
+    return(case)
+}
+
+## Check for "integers" but don't require true integer storage
+is.whole <- function (x) is.numeric(x) && floor(x) == x
+
+ensureValidCases <- function(cases, else_case) {
+    cases <- lapply(cases, ensureValidCase)
+
+    if (!missing(else_case)) {
+        # assign API default values if not given
+        defaults <- list(id=NULL, name=NULL, numeric_value=NULL, missing=FALSE)
+        else_case <- modifyList(defaults, else_case)
+
+        if(!is.null(else_case$expression)) {
+            halt("else_cases should not have any conditions expression")
+        }
+        if(!is.character(else_case$name)) {
+            halt("else_cases must have a (character) name")
+        }
+        if(!is.null(else_case$id) & !is.whole(else_case$id)) {
```

Comment:
This seems to duplicate `ensureValidCase`: why not run the "else_case" through there?